README.md
[![ObjectDetectionService](https://github.com/olahsymbo/object-detection-yolov5-fastapi/actions/workflows/object-detection-pipeline.yml/badge.svg)](https://github.com/olahsymbo/object-detection-yolov5-fastapi/actions/workflows/object-detection-pipeline.yml)
# Object Detection

A FastAPI object detection application based on Yolov5 model.

## Getting Started

- Clone the project
- `cd` into the codebase
- run `poetry shell` and `poetry install` to set the virtual environment and install the necessary dependencies

## Start the app
Since we used docker, we can start the app by running these commands:

```
sudo docker-compose build
sudo docker-compose up -d
```
The app will be served on port `8000`

## Make request

An example `curl` request:

``` 
curl --location '127.0.0.1:8000/object_detect' \
--form 'input_file=@"/Users/Downloads/download.jpeg"'
```

Valid response:
``` 
{
    "data": [
        [
            {
                "class": 0,
                "class_name": "person",
                "bbox": [
                    134,
                    330,
                    187,
                    579
                ],
                "confidence": 0.7723177671432495
            },
            {
                "class": 0,
                "class_name": "person",
                "bbox": [
                    90,
                    373,
                    140,
                    587
                ],
                "confidence": 0.7619432210922241
            },
            {
                "class": 2,
                "class_name": "car",
                "bbox": [
                    492,
                    314,
                    566,
                    430
                ],
                "confidence": 0.7516090273857117
            },
            {
                "class": 2,
                "class_name": "car",
                "bbox": [
                    379,
                    321,
                    455,
                    460
                ],
                "confidence": 0.7398971915245056
            },
            {
                "class": 0,
                "class_name": "person",
                "bbox": [
                    267,
                    332,
                    315,
                    541
                ],
                "confidence": 0.4890599846839905
            },
            {
                "class": 0,
                "class_name": "person",
                "bbox": [
                    592,
                    322,
                    611,
                    389
                ],
                "confidence": 0.45573562383651733
            },
            {
                "class": 2,
                "class_name": "car",
                "bbox": [
                    444,
                    313,
                    477,
                    400
                ],
                "confidence": 0.38912448287010193
            },
            {
                "class": 0,
                "class_name": "person",
                "bbox": [
                    209,
                    324,
                    260,
                    553
                ],
                "confidence": 0.30747783184051514
            }
        ]
    ],
    "message": "object detected successfully",
    "errors": null,
    "status": 200
}
```

Error response:

``` 
{
    "message": "object detection failed",
    "errors": "error",
    "status": 400
}
```


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.2.0 (2023-11-02)

### Feature

* Add git workflow ([`afd7134`](https://github.com/olahsymbo/object-detection-yolov5-fastapi/commit/afd7134dcbb6cf3a8d8888ec7803c7ecd886ba96))

## v0.1.0 (2023-11-02)

### Feature

* Add git workflow ([`3c03740`](https://github.com/olahsymbo/object-detection-yolov5-fastapi/commit/3c03740202b6e5531f26d461f1c9226fd6ff8e8c))
* Add git workflow ([`80f5755`](https://github.com/olahsymbo/object-detection-yolov5-fastapi/commit/80f5755e1a2235ff0e52bd9a9b5f4655691d1a34))
* Add git workflow ([`ee7c40b`](https://github.com/olahsymbo/object-detection-yolov5-fastapi/commit/ee7c40baeba2bf63b4418ff156be6743666d2eb3))
* Add git workflow ([`c7c18ac`](https://github.com/olahsymbo/object-detection-yolov5-fastapi/commit/c7c18ac6aa67c5177192e122c0cfc06ad3ac514f))
* Add git workflow ([`c831ae5`](https://github.com/olahsymbo/object-detection-yolov5-fastapi/commit/c831ae5b0fdfa7c2d53e00b1e7ea5ac515f3b283))
* **0.2.1:** Added yolo ([`ba4927f`](https://github.com/olahsymbo/object-detection-yolov5-fastapi/commit/ba4927ff58ac5799a767c4b450b9e5b16112c063))


apis/app.py
import os
import logging
from io import BytesIO
from typing import List
from warnings import filterwarnings, simplefilter
import ssl
import torch
import uvicorn
from fastapi import FastAPI, Request, File, UploadFile
from fastapi.responses import JSONResponse
from PIL import Image

from object_detection_yolov5.core.base_detector import ObjectDetector

try:
    _create_unverified_https_context = ssl._create_unverified_context
except AttributeError:
    pass
else:
    ssl._create_default_https_context = _create_unverified_https_context

filterwarnings("ignore")
simplefilter(action="ignore", category=FutureWarning)

if not os.path.exists("../logs"):
    os.mkdir("../logs")

logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)
logging.StreamHandler()
file_handler = logging.FileHandler("../logs/api.log")
file_handler.setLevel(logging.DEBUG)
formatter = logging.Formatter("%(asctime)s : %(levelname)s : %(name)s : %(message)s")
file_handler.setFormatter(formatter)
logger.addHandler(file_handler)

app: FastAPI = FastAPI()

model = torch.hub.load("ultralytics/yolov5", "yolov5l")


@app.post("/object_detect")
async def image_detect(request: Request, input_file: UploadFile = File(...)):

    if request.method == "POST":
        json_result: List = []
        try:

            image: Image = Image.open(BytesIO(await input_file.read()))
            ob: ObjectDetector = ObjectDetector(image, model)
            json_results = ob.object_detect()

            logger.info(["object detection results", json_result])

            return JSONResponse(
                {
                    "data": json_results,
                    "message": "object detected successfully",
                    "errors": None,
                    "status": 200,
                },
                status_code=200,
            )
        except Exception as error:
            logger.error(["process failed", error])
            return JSONResponse(
                {
                    "message": "object detection failed",
                    "errors": "error",
                    "status": 400,
                },
                status_code=400,
            )


if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)
